zpay32/fuzz_test.go

```diff
@@ -12,9 +13,44 @@ func FuzzDecode(f *testing.F) {
 	})
 }
 
+// appendChecksum returns a string containing bech followed by its bech32
+// checksum if a checksum could be calculated. Otherwise, the function returns
+// bech unchanged.
+func appendChecksum(bech string) string {
+	// This code is based on checksum calculation in zpay32/bech32.go.
+
+	lower := strings.ToLower(bech)
+
+	// The string is invalid if the last '1' is non-existent or it is the
+	// first character of the string (no human-readable part).
+	one := strings.LastIndexByte(lower, '1')
+	if one < 1 {
+		return bech
+	}
+	hrp := lower[:one]
+	data := lower[one+1:]
+
+	decoded, err := toBytes(data)
+	if err != nil {
+		return bech
+	}
+
+	checksum, err := toChars(bech32Checksum(hrp, decoded))
+	if err != nil {
+		return bech
+	}
+
+	return bech + checksum
+}
+
 func FuzzEncode(f *testing.F) {
 	f.Fuzz(func(t *testing.T, data string) {
-		inv, err := Decode(data, &chaincfg.TestNet3Params)
+		// Make it easier for the fuzzer to generate valid invoice
+		// encodings by adding the required prefix and valid checksum.
+		data = "lnbc" + data
```

Comment:
So the fuzzer still needs to find out that it needs to put a `1` into the data? But it probably will attempt to break things by placing the `1` anywhere in the data or even having multiple occurrences.

I guess it already did: https://github.com/lightninglabs/lnd-fuzz/blob/ae8685b553238aac3b6b87512f74ae33ab10a2a7/zpay32/testdata/fuzz/FuzzEncode/00e9523c69b9d9ee530ad43f0347fc96770752861eacd5f3c46c060b2d6677ab

Reply:
Yes, and the fuzzer seems to do well placing the `1` in different places.

The main reason I didn't manually add the `1` is so the fuzzer can generate the optional `amount` and `multiplier` in the HRP: https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#human-readable-part.

Reply:
I think its common for fuzzers to use certain special values like 0, 1, 255 more than others

Reply:
I think this should switch on the chaincfg chosen

Reply:
> I think this should switch on the chaincfg chosen

Done in 7390e6d23b6fd8329956a278892451ea6f0edb5f.